### PR TITLE
Classes and objects

### PR DESCRIPTION
**Pull request by MICHAELMUNAVU83**
- Implement classes and objects in Ruby.
- Implement encapsulation and inheritance with Ruby.
- Run a program using the command line.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    @parent_permission = parent_permission
+  end
+
+  private
+
+  def of_age
+    return true if @age >= 18
+
+    false
+  end
```
Comment (TedLivist) [inline person.rb]: - [ ] It is a Ruby convention to add the question mark (?) to methods of this nature that return `true` or `false` boolean values. Kindly change this method name to `of_age?`.

- [ ] **[OPTIONAL]** In keeping with DRY and KISS principles, we can write cleaner code here and save some code. Naturally, `@age >= 18` returns `true` or `false`, so we do not really need to write all this code just to return the same values. Kindly refactor to something like this:

  ```
  def of_age?
    @age >= 18
  end
  ```

File: person.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  def can_use_services
+    true if @age >= 18 || parent_permission == true
+  end
```
Comment (TedLivist) [inline person.rb]: - [ ] It is a Ruby convention to add the question mark (?) to methods of this nature that return true or false boolean values. Kindly change this method name to `can_use_services?`.

- [ ] We don't need to repeat ourselves here. We already have the `of_age?` method that evaluates if age is greater than or equal to 18. We can use that instead. We also know that the `@parent_permission` instance variable is either `true` or `false` so we don't need to check for it again. Kindly refactor this method to something like this:

  ```
  def can_use_services?
    of_age? || @parent_permission
  end
  ```

Comment (TedLivist) [inline person.rb]: - [ ] All methods under the `private` keyword are considered private, but only the `of_age?` method should be private. The can_use_services? should be public. Kindly remove it from among the private methods, like so:

  ```
  def can_use_services?
    ...
  end

  private

  def of_age?
    ...
  end
  ```

File: person.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class Person
+  attr_accessor :name, :age
+  attr_reader :id
+
+  def initialize(age, name = 'Unknown', parent_permission: true)
+    @id = id
```
Comment (TedLivist) [inline person.rb]: - [ ] You need to supply or auto-generate an `id` attribute. You can generate a random `id` like so:

  ```
  ...
  @id = Random.rand(1..1000)
  ...
  ```

File: student.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+require_relative 'person'
+class Student < Person
+  attr_accessor :classroom
+
+  def initialize(classroom, age, _name = 'Unknown', _parent_permission: true)
+    super(age, _name = 'Unknown', _parent_permission: true)
+    @classroom = classroom
+  end
```
Comment (TedLivist) [inline student.rb]: - [ ] Your `super` method prevents the passed-in parameters from being used in the object creation. After passing name and parent_permission arguments, your super method reverts everything to "Unknown", and `true` respectively.

- [ ] The underscore for does not seem to work well with the `_parent_permission` part. You probably need to refactor to something like so:
  ```
  def initialize(classroom, age, _name = 'Unknown', parent_permission: true)
    super(age, _name, parent_permission: parent_permission)
    @classroom = classroom
  end
  ```

File: teacher.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+require_relative 'person'
+class Teacher < Person
+  attr_accessor :specialization
+
+  def initialize(specialization, age, _name = 'Unknown')
+    super(age, _name = 'Unknown')
```
Comment (TedLivist) [inline teacher.rb]: - [ ] In your `super` method, ensure the passed in name is what is used if any, not "Unkown".

  ```
  super (age, _name)
  ```

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @MICHAELMUNAVU83,

Good job so far!
✔️ Good documentation
✔️ Classes are in different files
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: teacher.rb (hunk 5 above)
Comment (MICHAELMUNAVU83) [reply teacher.rb]: Thank you so much @TedLivist  for these corrections , You have taught me so much on ruby conventions . Thank you so so much .

**Review by dasileker — APPROVED** [review]
### **Hi @MICHAELMUNAVU83,** 
### **Nice work 👏👏**

### **Congratulations! 🎉**

Your project is complete!  There is nothing else to say other than... it's time to merge it 🚀 : 
 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**